### PR TITLE
[IMP] all: Viewtiverse: replace <tree> in <list>

### PR DESCRIPTION
**Pull request by Gorash**
Convert all "tree" views to "list" views.

Viewtiverse task-3599136
task-27709

### CODE REVIEW

**Comment by Gorash** [comment]
@mart-e With the others pr it's 7.660 changed lines.

The task is 7 years old, we should have done it earlier, there would have been fewer changes. The number of lines is not a relevant factor for changes. The change in this case is really trivial (that of the modifiers invisible/readonly/required was complicated, here it is a simple regexp).

Moreover, regarding changes such as views, Olivier suggest systems to facilitate fowardport/backport. We will have to see what we can put in place. This will make maintenance easier for many other trivial changes that currently exist.

**Comment by pedrobaeza** [comment]
I think this is not exactly an error, as when you group by any field, you have a "tree list", but anyway, the most important thing is the impact on all the downstream things: migrating any module to v18 means to change this to work as it was, and all the technical documentation and videos will be wrong for this new version. I have faced this recently with the removal of the `<act_window>` tag, and my new colleagues watching our training videos talking about this tag, and they are totally puzzled about the error installing the module.

Please reconsider it.

**Comment by mart-e** [comment]
> The number of lines is not a relevant factor for changes.

Yes it is, the more lines you change, the higher the chance to break an existing custo, to have a conflict during a forwardport,... As mentionned Pedro, it’s going to make the life of maintainer more difficult.
You have to take that into account when comparing with the benefit. I don’t say we should never do cleaning but only if it brings a noticable gain (which is not clear here).
I am sure we can find alternatives that doesn’t implies changing every occurrence of `tree`.

And if we haven’t done it in years, maybe it means it wasn’t that essential. A bit like renaming `res.users` to `res.user` or `stock` to `logistic`, probably never going to happen and that’s ok.

**Comment by Gorash** [comment]
@pedrobaeza  We can depreciate <tree> with a warning and in few year, <tree> will disappear. I don't think that it's a better solution for you. This is an old task which is controversial, several have worked on it previously for the spec. (The development was very rapid.) All opinions are welcome.

@mart-e For the forwardport, the merge-bot will take this change into account automatically (we hope it will be soon). The pr will surely be on the agenda of the next guru meeting. It will be interesting to see the current impact of the time of forward parts and causes.

Few months ago, I had merge the change of modifiers to have python expression (for invisible, required, read-only). With much more impact (I'm not saying that this is a reason to make now a lot of changes for nothing). It was to unify the beavour and the documentation. (For the previous big changes, some partners asked me for a script to automate their changes.)
In few time, every <kanban> view will be removed/updated. Every documentation of this views will changes. The goal is to have an easier api, and framework for the future.

In this case, it's a view changes (like all my actual tasks), their is a lot of visual change. In framework, every change can break the old api silent. In this case, the conflict can be automatically fixed.

Concerning other historical errors. Indeed, there are some that should be updated. Such as groups_id or others.

**Comment by pedrobaeza** [comment]
Hi, Cristophe, deploying a depreciation strategy may help a bit, but at the end, the moment it's removed, we'll have the same problem. And even in between, the cherry-picks between versions will also be complicated with no gain except being a bit more correct in the terminology.

Although you are handling the change on your fw-port bot, think in the rest of the developers that depend on your framework and don't have such automation possibilities. For example, in OCA, we can't deploy a fw-port bot, as the stability policies can't be as strict as yours (the same as the migration strategy), and the conflict possibilities are higher, and OCA developers don't have direct access to the branches to solve possible conflicts, so it doesn't worth to put it for most of the times not being able to merge.

**Comment by carlosecv** [comment]
While changing the tags from “tree” to “list” may seem like a minor adjustment at first glance, it’s essential to consider the potential impact on the broader Odoo community. This type of modification could lead to confusion and necessitate users to adjust their existing practices and knowledge. Furthermore, it may introduce compatibility issues with existing modules, requiring significant time and effort to resolve.

It’s crucial to carefully weigh the potential benefits of this change against the associated costs, both in terms of the community workload and the overall consistency and usability of the system. In many cases, maintaining consistency with existing conventions may be more prudent unless there’s a compelling reason to justify the change.

**Comment by ged-odoo** [comment]
guys, this task comes from Antony. go talk to him if you want, but please leave Christophe in peace

**Comment by pedrobaeza** [comment]
I think the tone from now has been correct and only stating the facts and highlighting the pain/gain ratio. Sorry if my comments have annoyed anyone. Please tell Antony to read these comments to reconsider the task.

**Comment by jakubsmolka** [comment]
I am actually rooting for Odoo here. Yes, it does make upgrade of customizations **slightly** more complex (replacing tree with list does not really seem difficult, it is only about number of occurences and automation is possible). And I am always in favour of keeping things clean. Trees were only confusing newcomers to Odoo technical world - however miniscule the change is, the underlying principle is correct and worth following in my opinion.

**Comment by gaikaz** [comment]
I must agree that the change seems trivial, but tend to agree with doing it sooner than later.
Isn't this exactly why Odoo chose major version releases every year so that breaking changes like this could happen?

**Comment by ossimantylahti** [comment]
Looking at this change makes me wonder what is the benefit from breaking so many old implementations in O18. Just to clean up the style? Or are we planning on using the 'tree' for something else?

The pain is relatively simple to estimate: for many many extensions the same code has been working between 14, 15, 16 and 17. This change will break the compatibility without any clear benefit. I would reconsider this quite carefully.

**Comment by rvalyi** [comment]
I'm disappointed the branch is not named something-pocalypse. Traditions seems to be forgotten at Odoo, yes apocalypse PRs were better before. Just in case I suggest to name the branch apocalist :-D

**Comment by randall-vx** [comment]
@nhomar @desdelinux @luisg123v @moylop260 

image

**Comment by nhomar** [comment]
Go ahead... worst things happens in the past and this makes totally sense!!! It is a list not a tree... let's go and do a regex find and replace in our community branches.

**Comment by mart-e** [comment]
Locking until decision has been taken as I believe the arguments have been stated. Thanks to keep things calm and avoid unnecessary "I agree" comments.

**Comment by Gorash** [comment]
@ossimantylahti Indeed, we would like to free the tag in order to be able to create real `tree` views later (there is no spec for this yet, but the idea is present).

NB: The upgrade between v16 and V17 also requires changes because the semantics of attributes such as `invisible`, `readonly`, `required`, `attrs` or `states` have been totally modified (it's now python expression instead of domain and others syntax, it was an other Viewtiverse tasks).

As @mart-e says, the arguments will be analyzed at our next guru meeting.

File: odoo/cli/code_upgrade.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        module.upgrade(file_manager)
+
+    for file in file_manager:
+        if file.dirty:
+            print('updated: ', file.path)  # noqa: T201
+            if not test:
+                file.path.open("w").write(file.content)
+
+
+class Code_Upgrade(Command):
```
Comment (Julien00859) [inline odoo/cli/code_upgrade.py:102]:
```py
class CodeUpgrade(Command):
    name = 'code_upgrade'

File: odoo/cli/code_upgrade.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    def run(self, cmdargs):
+        odoo = Path(__file__).parent.parent.parent
+        to_version = SourceFileLoader('release', str(next(odoo.glob('odoo/release.py')))).load_module().version_info
```
Comment (kmagusiak) [inline odoo/cli/code_upgrade.py:105]: You can just import `odoo.release`.

File: odoo/code_upgrade/17.5-00-tree-to-list.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import re
```
Comment (kmagusiak) [inline odoo/code_upgrade/17.5-00-tree-to-list.py:1]: Shouldn't code upgrades be inside `odoo.upgrades`?

Comment (Julien00859) [reply odoo/code_upgrade/17.5-00-tree-to-list.py:1]: no we want it inside regular odoo/

Comment (kmagusiak) [reply odoo/code_upgrade/17.5-00-tree-to-list.py:1]: Maybe rename to `odoo/upgrade_code` so that it is besides the other package?

**Comment by Julien00859** [comment]
Also when we invoke the command without any `files` I suggest we walk through all addons paths. `for ad_path in odoo.addons.__path__: for ... in os.walk(ad_path): ...`

**Comment by Xavier-Do** [comment]
> The FileDescriptor FileManager can be simplified if we use the functions from `odoo.modules.modules`, the `initialize_sys_path`, `get_module_root`, `get_manifest`. We can also use the `odoo.modules.graph` module to do the module dependency tree.

We want to be able to execute this script in a very minimal and cross version environment, without the odoo dependencies. Thus, even if the command line is present, the script *may* be used to migrate diffs, executed in runbot to suggest forwardport changes in case of conflict, migrate customers modules, ... One of the specification was to only need basic imports like re, lxml, ... So we shouldn't be dependant of odoo tools. 


> I suggest to end up with the following API for code migration scripts:
> 
> ```python
> def upgrade(paths):
>     for path in paths:
>         with open(path) as file:
>             content = file.read()
> 
>         ...  # code-upgrade code comes here
> 
>         with open(path, 'w') as file:
>             file.write(content)
> ```
> 
> A bit more of boilerplates but ultimately simpler than learning with APIs (FileDescription, FileManager).

The api is there to allow to easily give something else than a list of file on the object (modules, static xml files, ....) 
The fact that we don't have an open will allow to execute dry run, mainly on readonly file-system where we just want to output de expected diff, or generate the commit using an api without writing on files


> Let's also change the `--glob` for a `add_argument('files', nargs='*')`. I mean the shell is gonna resolves the glob itself in most cases...

You would have to list all files from all addons_path? 
The goal of this glob is to filter out some files, mainly targeting the case where we want to apply it on a specific file that had some diff. But having the addons_path as an api will be useful. With a glob, we may miss the information on what is a module, or not, a static file or not, a data or not. We need to be able to qualify precisely a file to know how to migrate it. It is not doable with a raw list of files.

File: odoo/cli/code_upgrade.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+"""
+Script method is add into the /upgrade/code directory.
+Every script should be in a file corresponding to his upgrade version (script for the )
+
+"""
+
+AVAILABLE_EXT = ('py', 'js', 'css', 'scss', 'xml', 'csv')
+
+
+class FileDescriptor:
```
Comment (rco-odoo) [inline odoo/cli/code_upgrade.py:21]: The purpose of this class is to provide some abstraction for getting and setting a file's content, possibly without writing it for real.  So it's not a file _descriptor_, which refers to a specific OS object.  I suggest to call this a `FileAccessor`.

File: odoo/cli/code_upgrade.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    def __init__(self, path: Path, addon: Path) -> None:
+        self.path = path
+        self.addon = addon
+        self._content = None
+        self.dirty = False
+
+    @property
+    def content(self):
+        if self._content is None:
+            self._content = self.path.open("r").read()
```
Comment (rco-odoo) [inline odoo/cli/code_upgrade.py:35]:
```suggestion
            with self.path.open("r") as f:
                self._content = f.read()
```

File: odoo/cli/code_upgrade.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+    def __init__(self, addons_path: list[str], glob: str = '**/*') -> None:
+        self.addons_path = addons_path
+        self.glob = glob
+        self._files = {}
+        for addon_path in addons_path:
+            for path_addon in Path(addon_path).iterdir():
+                for path in path_addon.glob(glob):
+                    if path.is_file() and path.name.rsplit('.')[-1] in AVAILABLE_EXT:
+                        self._files[str(path)] = FileDescriptor(path, path_addon)
```
Comment (rco-odoo) [inline odoo/cli/code_upgrade.py:57]:
```suggestion
        self._files = {
            str(path): FileDescriptor(path, path_addon)
            for addon_path in addons_path
            for addon in Path(addon_path).iterdir()
            for path in addon.glob(glob)
            if path.is_file() and path.suffix in AVAILABLE_EXT
        }
```

File: odoo/cli/code_upgrade.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    file_manager = FileManager(addons_path, glob)
+    for (name, module) in modules:
+        print(f'update script: {name}')  # noqa: T201
+        module.upgrade(file_manager)
+
+    for file in file_manager:
+        if file.dirty:
+            print('updated: ', file.path)  # noqa: T201
+            if not test:
+                file.path.open("w").write(file.content)
```
Comment (rco-odoo) [inline odoo/cli/code_upgrade.py:99]:
```suggestion
                with file.path.open("w") as f:
                    f.write(file.content)
```

File: odoo/addons/base/models/ir_ui_view.py [hunk 8]
```diff
@@ -520,6 +520,12 @@ def write(self, vals):
         if 'arch_updated' not in vals and ('arch' in vals or 'arch_base' in vals) and 'install_filename' not in self._context:
             vals['arch_updated'] = True
 
+        if 'arch_db' in vals and self.type not in self._fields['type']._selection:
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_ui_view.py:523]: This looks a lot like you are hiding some bug.  Why do you need this?

**Review by rco-odoo — COMMENTED** [review]
The page is so slow that I could not review it properly.
I only managed to check what changed in the subdirectory "odoo"...

File: odoo/addons/base/models/ir_ui_view.py (hunk 8 above)
Comment (rco-odoo) [reply odoo/addons/base/models/ir_ui_view.py:523]: Moreover, `self` may be a recordset with more than one record.  In that case, `self.type` will crash.

File: odoo/cli/code_upgrade.py (hunk 2 above)
Comment (Gorash) [reply odoo/cli/code_upgrade.py]: we don't want to import odoo into this file

**Comment by rco-odoo** [comment]
@robodoo rebase-ff

File: odoo/cli/upgrade_code.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+    def __init__(self, addons_path: list[str], glob: str = '**/*') -> None:
+        self.addons_path = addons_path
+        self.glob = glob
+        self._files = {
+            str(path): FileAccessor(path, addon)
+            for addon_path in addons_path
+            for addon in Path(addon_path).iterdir()
+            for path in addon.glob(glob)
+            if path.is_file() and path.name.rsplit('.')[-1] in AVAILABLE_EXT
```
Comment (rco-odoo) [inline odoo/cli/upgrade_code.py:59]: There's an API for getting a file extension from its `Path`:
```suggestion
            if path.is_file() and path.suffix in AVAILABLE_EXT
```

File: odoo/upgrade_code/17.5-00-tree-to-list.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    reg_tree_to_list_ref = re.compile(r'tree_view_ref')
+    reg_tree_to_list_mode = re.compile(r"""(mode=['"][^'"]*)tree([^'"]*['"])""")
+    reg_tree_to_list_view_mode = re.compile(r"""(['"]view_mode['"][^'":=]*[:=].*['"]([^'"]+,)?\s*)tree(\s*(,[^'"]+)?['"])""")
+    reg_tree_to_list_view = re.compile(r"""(['"]views['"][^'":]*[:=].*['"])tree(['"])""")
+    reg_tree_to_list_string = re.compile(r"""([ '">)])tree( [vV]iews?[ '"<.)])""")
+    reg_tree_to_list_String = re.compile(r"""([ '">)])Tree( [vV]iews?[ '"<.)])""")
+    reg_tree_to_list_env_ref = re.compile(r"""(self\.env\.ref\(.*['"])tree(['"])""")
+
+    for file in file_manager:
+        if file.path.name.rsplit('.', 1)[1] in ('xml', 'js', 'py'):
```
Comment (rco-odoo) [inline odoo/upgrade_code/17.5-00-tree-to-list.py:21]:
```suggestion
        if file.path.suffix in ('xml', 'js', 'py'):
```

File: odoo/upgrade_code/17.5-00-tree-to-list.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+            content = reg_tree_to_list_ref.sub('list_view_ref', content)
+            content = reg_tree_to_list_mode.sub(r'\1list\2', content)
+            content = reg_tree_to_list_view_mode.sub(r'\1list\3', content)
+            content = reg_tree_to_list_view.sub(r'\1list\2', content)
+            content = reg_tree_to_list_string.sub(r'\1list\2', content)
+            content = reg_tree_to_list_String.sub(r'\1List\2', content)
+            content = reg_tree_to_list_env_ref.sub(r'\1list\2', content)
+            file.content = content
+
+            i += 1
```
Comment (rco-odoo) [inline odoo/upgrade_code/17.5-00-tree-to-list.py:36]: There's a Python idiom for maintaining an index in a loop:
```py
file_manager.print_progress(index + 1, total)
for index, element in enumerate(elements):
    ...
    file_manager.print_progress(index + 1, total)
```
Or this one is even more elegant:
```py
for index, element in enumerate(elements):
    file_manager.print_progress(index, total)
    ...
file_manager.print_progress(total, total)
```

File: odoo/upgrade_code/17.5-00-tree-to-list.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
+import re
+
+
+def upgrade(file_manager):
+    nb = len([1 for file in file_manager if file.path.name.rsplit('.', 1)[1] in ('xml', 'js', 'py')])
```
Comment (rco-odoo) [inline odoo/upgrade_code/17.5-00-tree-to-list.py:5]: `nb` does not mean anything, except it's a number.  At least `total` suggests to total amount:
```suggestion
    total = len([1 for file in file_manager if file.path.suffix in ('xml', 'js', 'py')])
    if not total:
        return  # nothing to do!
```

**Review by rco-odoo — COMMENTED** [review]
Your upgrade script will crash with a division by zero if you use a glob and there's no file to process.

**Comment by rco-odoo** [comment]
@robodoo delegate+

**Comment by Gorash** [comment]
@robodoo r+ alone

**Comment by d-fence** [comment]
robodoo r+ alone

**Pull request closed without merging**